src/internal/utils/date-time/calendar.ts

```diff
@@ -0,0 +1,165 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import addMonths from 'date-fns/addMonths';
+import isAfter from 'date-fns/isAfter';
+import isBefore from 'date-fns/isBefore';
+import isSameDay from 'date-fns/isSameDay';
+import isSameMonth from 'date-fns/isSameMonth';
+import subMonths from 'date-fns/subMonths';
+import { getCalendarMonth } from 'mnth';
+
+import { DayIndex } from '../locale';
+
+// Returns a 3-by-4 matrix with dates corresponding to the initial date-time of each month of the year for a given date.
+export function getCalendarYear(date: Date): Date[][] {
+  const year = date.getFullYear();
+  return new Array(4)
+    .fill(0)
+    .map((_, i: number) => new Array(3).fill(0).map((_, j: number) => new Date(year, i * 3 + j)));
+}
+
+interface CalendarWeek {
+  days: CalendarDate[];
+  testIndex?: number;
+}
+
+interface CalendarDate {
+  date: Date;
+  isVisible: boolean;
+  isSelected: boolean;
+  isInRange: boolean;
+  isSelectionTop: boolean;
+  isSelectionBottom: boolean;
+  isSelectionLeft: boolean;
+  isSelectionRight: boolean;
+}
+
+export class MonthCalendar {
```

Comment:
The calendar abstraction helps calculating indices and styles for calendar grid, which is non-trivial due to presence of placeholder dates that are treated differently.

Ideally, the same abstraction should be reused between Calendar, and Date range picker components. However, those are currently using very different implementations and only one supports the year calendar. We can reuse the code between the components as soon we update those to use the same calendar element implementation.